test/metabase/db/custom_migrations_test.clj

```diff
@@ -332,3 +333,130 @@
+                        {:row 36 :col 0  :size_x 17 :size_y 1}
+                        {:row 36 :col 17 :size_x 1  :size_y 1}]
+          revision-id (first (t2/insert-returning-pks! 'Revision
+                                                        {:object   {:cards cards}
+                                                         :model    "Dashboard"
+                                                         :model_id 1
+                                                         :user_id  user-id}))]
+
+      (migrate!)
+      (testing "forward migration migrate correclty"
+        (is (= [{:col 0  :row 20 :size_x 16 :size_y 10}
+                {:col 16 :row 9  :size_x 8  :size_y 11}
+                {:col 7  :row 2  :size_x 6  :size_y 4}
+                {:col 0  :row 33 :size_x 9  :size_y 13}
+                {:col 0  :row 2  :size_x 7  :size_y 4}
+                {:col 8  :row 9  :size_x 8  :size_y 11}
+                {:col 9  :row 33 :size_x 15 :size_y 13}
+                {:col 0  :row 9  :size_x 8  :size_y 5}
+                {:col 0  :row 30 :size_x 24 :size_y 3}
+                {:col 0  :row 6  :size_x 24 :size_y 3}
+                {:col 0  :row 0  :size_x 24 :size_y 2}
+                {:row 48 :col 0  :size_x 23 :size_y 1}
+                {:row 48 :col 23 :size_x 1  :size_y 1}]
+               (t2/select-one-fn (comp :cards :object) :model/Revision :id revision-id))))
+     (migrate-down! 46)
+     (testing "downgrade works correctly"
+      (is (= [{:col 0  :row 15 :size_x 12 :size_y 8}
+              {:col 12 :row 7  :size_x 6  :size_y 8}
+              {:col 5  :row 2  :size_x 4  :size_y 3}
+              {:col 0  :row 25 :size_x 7  :size_y 10}
+              {:col 0  :row 2  :size_x 5  :size_y 3}
+              {:col 6  :row 7  :size_x 6  :size_y 8}
+              {:col 7  :row 25 :size_x 11 :size_y 10}
+              {:col 0  :row 7  :size_x 6  :size_y 4}
+              {:col 0  :row 23 :size_x 18 :size_y 2}
+              {:col 0  :row 5  :size_x 18 :size_y 2}
+              {:col 0  :row 0  :size_x 18 :size_y 2}
+              {:row 36 :col 0  :size_x 17 :size_y 1}
+              {:row 36 :col 17 :size_x 1  :size_y 1}]
+           (t2/select-one-fn (comp :cards :object) :model/Revision :id revision-id)))))))
+
+(defn two-cards-overlap? [box1 box2]
+  (let [{col1    :col
+         row1    :row
+         size_x1 :size_x
+         size_y1 :size_y} box1
+        {col2    :col
+         row2    :row
+         size_x2 :size_x
+         size_y2 :size_y} box2]
+    (and (< col1 (+ col2 size_x2))
+         (> (+ col1 size_x1) col2)
+         (< row1 (+ row2 size_y2))
+         (> (+ row1 size_y1) row2))))
+
+(defn no-cards-are-overlaped?
+  "Return false if the cards contains at least 1 pair of cards that overlap, else returns true"
+  [boxes]
+  (every? false? (for [i (range (count boxes))
+                        j (range (- (count boxes) i 1))]
+                   (two-cards-overlap? (nth boxes i) (nth boxes (+ i j 1))))))
```

Comment:
Took me a while to understand this was just combinations of boxes. This might be more readable:
```suggestion
(not (some two-cards-overlap? (clojure.math.combinatorics/combinations boxes 2)))
```

Reply:
nice! 4f169ca09c